perses/dispersed/feptasks.py

```diff
@@ -84,30 +111,103 @@ def _get_integrator(self, thermodynamic_state):
         """
         return self._integrator
 
-    def reset(self):
-        """
-        Reset the work statistics on the associated ContextCache integrator.
+    def apply(self, thermodynamic_state, sampler_state):
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It will apply the full NCMC protocol.
 
         Parameters
         ----------
         thermodynamic_state : openmmtools.states.ThermodynamicState
-            the thermodynamic state for which this integrator is cached.
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
         """
-        self._integrator.reset()
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It also logs
+        benchmarking information through the utils.Timer class.
+        Parameters
+        ----------
+        thermodynamic_state : openmmtools.states.ThermodynamicState
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
+        See Also
+        --------
+        openmmtools.utils.Timer
+        """
+        # Check if we have to use the global cache.
+        if self.context_cache is None:
+            context_cache = cache.global_context_cache
+        else:
+            context_cache = self.context_cache
+
+        # Create integrator.
+        integrator = self._get_integrator(thermodynamic_state)
+
+        context, integrator = context_cache.get_context(thermodynamic_state, integrator)
+
+        # Subclasses may implement _before_integration().
+        self._before_integration(context, thermodynamic_state)
+        
+        self._cumulative_work[0] = integrator.get_protocol_work()
+
+        if self._cumulative_work != 0.0:
+            raise RuntimeError("The initial cumulative work after reset was not zero.")
+
+        #loop through the number of times we have to apply in order to collect the requested work and trajectory statistics.
+        for iteration in range(self._number_of_step_moves):
+
+            integrator.step(self._work_save_interval)
+            self._current_total_work = integrator.get_protocol_work(dimensionless=True)
+            self._cumulative_work[iteration+1] = self._current_total_work
+
+            #if we have a trajectory, we'll also write to it
+            if self._topology is not None:
+                context_state = context.getState(getPositions=True, getVelocities=True, getEnergy=True,
+                                                 enforcePeriodicBox=thermodynamic_state.is_periodic)
+                sampler_state.update_from_context(context_state)
```

Comment:
You can safely pass the `context` here instead of the OpenMM `State`. `SamplerState` will take care of getting only what it needs and cache the unit-less positions/velocities in the meantime so you'll get some speedup this way.

The only reason why in the original `apply()` method we pass the `State` object is because we need the energy as part of the NaN restarting code.

Reply:
ah ok, thanks